api/v1/server/handlers/v1/webhooks/receive.go

```diff
@@ -239,11 +254,104 @@ func (vr ValidationError) ToResponse() (gen.V1WebhookReceiveResponseObject, erro
 	}
 }
 
+type IsValid bool
+type IsChallenge bool
+
+func (w *V1WebhooksService) performChallenge(webhookPayload []byte, webhook sqlcv1.V1IncomingWebhook, request http.Request) (IsChallenge, map[string]interface{}, error) {
+	switch webhook.SourceName {
+	case sqlcv1.V1IncomingWebhookSourceNameSLACK:
+		payload := make(map[string]interface{})
+		err := json.Unmarshal(webhookPayload, &payload)
+
+		if err != nil {
+			return false, nil, fmt.Errorf("failed to parse form data: %s", err)
+		}
+
+		if challenge, ok := payload["challenge"].(string); ok && challenge != "" {
+			return true, map[string]interface{}{
+				"challenge": challenge,
+			}, nil
+		}
+
+		return false, nil, nil
+	default:
+		return false, nil, nil
+	}
+}
+
 func (w *V1WebhooksService) validateWebhook(webhookPayload []byte, webhook sqlcv1.V1IncomingWebhook, request http.Request) (
-	bool,
+	IsValid,
 	*ValidationError,
 ) {
 	switch webhook.SourceName {
+	case sqlcv1.V1IncomingWebhookSourceNameSLACK:
+		timestampHeader := request.Header.Get("X-Slack-Request-Timestamp")
+
+		if timestampHeader == "" {
+			return false, &ValidationError{
+				Code:      Http403,
+				ErrorText: "missing or invalid timestamp header: X-Slack-Request-Timestamp",
+			}
+		}
+
+		timestamp, err := strconv.ParseInt(strings.TrimSpace(timestampHeader), 10, 64)
+
+		if err != nil {
+			return false, &ValidationError{
+				Code:      Http403,
+				ErrorText: fmt.Sprintf("invalid timestamp in header: %s", err),
+			}
+		}
+
+		// qq: should this be utc?
+		if time.Unix(timestamp, 0).UTC().Before(time.Now().Add(-5 * time.Minute)) {
+			return false, &ValidationError{
+				Code:      Http403,
+				ErrorText: "timestamp in header is out of range",
+			}
+		}
+
+		algorithm := webhook.AuthHmacAlgorithm.V1IncomingWebhookHmacAlgorithm
+		encoding := webhook.AuthHmacEncoding.V1IncomingWebhookHmacEncoding
+		decryptedSigningSecret, err := w.config.Encryption.Decrypt(webhook.AuthHmacWebhookSigningSecret, "v1_webhook_hmac_signing_secret")
+
+		if err != nil {
+			return false, &ValidationError{
+				Code:      Http500,
```

Comment:
Should this be a 500 response?

Reply:
I think so? same for the one below too (failing to compute the signature, although maybe that one is more debatable and could be e.g. 403)

Reply:
ah yeah sorry, we will fail and return 403 if the signatures don't match, but that error is just checking if we fail to compute the signature, not if it doesn't match the provided one

Reply:
got it, seems reasonable then